code.tsx

```diff
@@ -96,13 +96,24 @@ function chooseTextColor (backgroundColor: string): WidgetJSX.Color {
 }
 
 function DatabaseTableWidget() {
-  const [theme, setTheme] = useSyncedState("theme", colors[0].option)
+  const [color, setColor] = useSyncedState("theme", colors[0].option)
   const [tableName, setTableName] = useSyncedState("tableName", null)
   const [columns, setColumns] = useSyncedState("columns", (): Column[] => [])
 
   useEffect(() => {
-    figma.ui.onmessage = (newColumns) => {
-      setColumns(JSON.parse(newColumns))
+    figma.ui.onmessage = (message) => {
+      const { action, data } = JSON.parse(message)
+      switch (action) {
+        case 'edit':
+          setColumns(data)
+          break
+        case 'raw-edit':
+          setColumns(data.columns)
+          setColor(data.color)
+          setTableName(data.tableName)
```

Comment:
Sharing my thoughts

I don't like this, because it wil trigger a render 3 times.
I was going to replace single properties with a single "tableState" state, but it would break existing users 😕

Maybe adding a code to first sync old with the new state and then removing would better help in transitioning users, but if someone missed the update, the table might be broken on another update

Due to this chance of breaking users, I decided to let them the same 🤷

What do you think?

Reply:
I believe we could change the States to Maps, but it didn't cancel the 3 time rendering.

If JSX and FigJam are smart enough, and they trigger the render via `requestAnimationFrame` (which I believe it does), it simply doesn't matter, and the framework only do the rendering one time.